drivers/axi_core/spi_engine/spi_engine.c

```diff
@@ -0,0 +1,958 @@
+				    SPI_ENGINE_CMD_CONFIG(
+					SPI_ENGINE_CMD_REG_CONFIG,
+					desc->mode));
+
+	/* Configure the prescaler */
+	spi_engine_queue_append_cmd(&msg->cmds,
+				    SPI_ENGINE_CMD_CONFIG(
+					SPI_ENGINE_CMD_REG_CLK_DIV,
+					desc_extra->clk_div));
+
+	/* Add a sync command to signal that the transfer has finished */
+	spi_engine_queue_add_cmd(&msg->cmds, SPI_ENGINE_CMD_SYNC(_sync_id));
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initiate a spi transfer
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param msg Structure used to store the transfer messages
+ * @return int32_t - SUCCESS if the transfer finished
+ *		   - FAILURE if the memory allocation failed
+ */
+int32_t spi_engine_transfer_message(struct spi_desc *desc,
+				    struct spi_engine_msg *msg)
+{
+	uint32_t		i;
+	uint32_t		data;
+	uint32_t		sync_id;
+	bool 			offload_en;
+	struct spi_engine_desc	*desc_extra;
+
+	desc_extra = desc->extra;
+
+	spi_engine_compile_message(desc, msg);
+
+	offload_en = (desc_extra->offload_config & OFFLOAD_TX_EN) |
+		     (desc_extra->offload_config & OFFLOAD_RX_EN);
+
+	/* Write the command fifo buffer */
+	do {
+		spi_engine_queue_lifo_get_cmd(&msg->cmds, &data);
+		spi_engine_write_cmd(desc, data);
+		usleep(1000);
+	} while(msg->cmds->next != NULL);
+	/* Write the last command */
+	spi_engine_queue_fifo_get_cmd(&msg->cmds, &data);
+	spi_engine_write_cmd(desc, data);
+
+	/* Write a number of tx_length WORDS on the SDO line */
+
+	if(offload_en) {
+		for(i = 0; i < desc_extra->offload_tx_len; i++) {
+			spi_engine_write(desc_extra,
+					 SPI_ENGINE_REG_OFFLOAD_SDO_MEM(0),
+					 bswap_constant_32(msg->tx_buf[i]));
+		}
+
+	} else {
+		for(i = 0; i < msg->length; i++) {
+			spi_engine_write(desc_extra,
+					 SPI_ENGINE_REG_SDO_DATA_FIFO,
+					 msg->tx_buf[i]);
+		}
+		do {
+			spi_engine_read(desc_extra,
+					SPI_ENGINE_REG_SYNC_ID,
+					&sync_id);
+		}
+		/* Wait for the end sync signal */
+		while(sync_id != _sync_id);
+		_sync_id++;
+
+		/* Read a number of rx_length WORDS from the SDI line and store
+		them */
+		for(i = 0; i < msg->length; i++) {
+			spi_engine_read(desc_extra,
+					SPI_ENGINE_REG_SDI_DATA_FIFO,
+					&data);
+			msg->rx_buf[i] = data;
+		}
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the spi engine
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param param Structure containing the spi init parameters
+ * @return int32_t - SUCCESS if the transfer finished
+ *		   - FAILURE if the memory allocation failed
+ */
+int32_t spi_engine_init(struct spi_desc **desc,
+			const struct spi_init_param *param)
+{
+	uint32_t			data_width;
+	uint32_t			spi_engine_version;
+	struct spi_desc			*local_desc;
+	struct spi_engine_desc		*eng_desc;
+	struct spi_engine_init_param	*spi_engine_init;
+
+	local_desc = (struct spi_desc *)malloc(sizeof(*local_desc));
+	eng_desc = (struct spi_engine_desc*)malloc(sizeof(*eng_desc));
+
+	if (!desc || !eng_desc)
+		return FAILURE;
```

Comment:
In my opinion, the check after each allocation seems to be redundant since in the application you free (remove) the descriptor in case of init failure.